### PR TITLE
Converted model classes to records

### PR DESCRIPTION
**Pull request by danjagnow**
Updated ArgentPonyWarcraftClient.csproj to set LangVersion to latest to enable the usage of records.  Wrote a crude script to automate the conversion of all model classes to records.

### CODE REVIEW

**Comment by danjagnow** [comment]
For reference, here is the very crude conversion script:

```cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace ClassToRecordThing
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            //ConvertToRecord(@"D:\Dev\warcraft\src\ArgentPonyWarcraftClient\Models\GameDataApi\Achievement\Achievement.cs");

            foreach (string path in Directory.EnumerateFiles(@"D:\Dev\warcraft\src\ArgentPonyWarcraftClient\Models", "*.cs", SearchOption.AllDirectories))
            {
                ConvertToRecord(path);
                Console.WriteLine(path);
            }

            Console.WriteLine("Done!");
        }

        private static void ConvertToRecord(string path)
        {
            State state = State.BeforeClass;
            string className = null;
            string currentSummary = null;
            var properties = new List<PropertyDetails>();

            StringBuilder final = new StringBuilder();

            foreach(string line in File.ReadAllLines(path))
            {
                string outline = line;

                switch(line) {
                    case string c when state == State.BeforeClass && c.Contains("public class"):
                        state = State.InClass;
                        className = Regex.Match(c, @"public class (\w+)").Groups[1].Value;
                        outline = c.Replace("public class", "public record");
                        break;
                    case string c when state == State.InClass && c.Contains("/// <summary>"):
                        state = State.InSummary;
                        break;
                    case string c when state == State.InSummary && c.Contains("/// Gets "):
                        currentSummary = Regex.Match(c, @"/// Gets (.+)$").Groups[1].Value;
                        break;
                    case string c when state == State.InSummary && c.Contains("/// </summary>"):
                        state = State.InClass;
                        break;
                    case string c when state == State.InClass && c.Contains("public ") && c.Contains("{ get; "):
                        Match match = Regex.Match(c, @"public ([^\s]+) (\w+) { get;");
                        string currentType = match.Groups[1].Value;
                        string currentProperty = match.Groups[2].Value;
                        properties.Add(new PropertyDetails { Type = currentType, NameUpper = currentProperty, NameCamel = LowerFirst(currentProperty), Summary = currentSummary });
                        outline = c.Replace("{ get; set; }", "{ get; }");
                        break;
                    case string c when state == State.InClass && c.Contains("    }"):
                        state = State.OutClass;
                        final.Append(BuildConstructor(className, properties));
                        break;
                }

                final.AppendLine(outline);
            }

            File.WriteAllText(path, final.ToString());
        }

        private static string BuildConstructor(string className, List<PropertyDetails> properties)
        {
            var c = new StringBuilder();

            c.AppendLine("");
            c.AppendLine("        /// <summary>");
            c.AppendLine($@"        /// Initializes a new instance of the <see cref=""{className}""/> class.");
            c.AppendLine("        /// </summary>");

            foreach(PropertyDetails property in properties)
            {
                c.AppendLine($@"        /// <param name=""{property.NameCamel}"">{UpperFirst(property.Summary)}</param>");
            }

            c.AppendLine("        [JsonConstructor]");
            string propertyList = string.Join(", ", properties.Select(p => $@"{p.Type} {p.NameCamel}"));
            c.AppendLine($@"        public {className}({propertyList})");

            c.AppendLine("        {");

            foreach (PropertyDetails property in properties)
            {
                c.AppendLine($@"            {property.NameUpper} = {property.NameCamel};");
            }

            c.AppendLine("        }");

            return c.ToString();
        }

        private static string LowerFirst(string text)
        {
            if (text != null && text.Length > 0)
            {
                return char.ToLower(text[0]) + text.Substring(1);
            }

            return text;
        }

        private static string UpperFirst(string text)
        {
            if (text != null && text.Length > 0)
            {
                return char.ToUpper(text[0]) + text.Substring(1);
            }

            return text;
        }

        private enum State
        {
            BeforeClass = 0,
            InClass = 1,
            InSummary = 2,
            OutClass = 3
        }

        private class PropertyDetails
        {
            public string NameUpper { get; set; }
            public string NameCamel { get; set; }
            public string Type { get; set; }
            public string Summary { get; set; }
        }
    }
}

```

**Review by willwolfram18 — APPROVED** [review]
Looks good and makes sense to me! One thing I do want to call out, which you may have already thought about, is that this is going to require a major version bump on the library since the `set` on properties has been removed. It's also worth highlighting that by removing the setters and moving to a constructor, it does make it harder for consumers to write any unit tests against the library. Not impossible, just mildly more inconvenient since they have to call the constructor with multiple parameters.

**Comment by danjagnow** [comment]
> One thing I do want to call out, which you may have already thought about, is that this is going to require a major version bump on the library since the `set` on properties has been removed.

Yes, the next release was already going to require a major version bump due to the changes in #183.  There weren't a lot of changes, but they were breaking changes.  Might as well take advantage of it.

> It's also worth highlighting that by removing the setters and moving to a constructor, it does make it harder for consumers to write any unit tests against the library. Not impossible, just mildly more inconvenient since they have to call the constructor with multiple parameters.

That's a very good point.  I reverted the change and took the approach of making the properties init-only instead of adding constructors.  That's probably more readable.

File: src/ArgentPonyWarcraftClient/IsExternalInit.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿#if NETSTANDARD2_0
+
+namespace System.Runtime.CompilerServices
+{
+    // This is required to enable the init keyword while targeting lower than .NET 5
+    // See https://stackoverflow.com/a/64749403/342581.
```
Comment (willwolfram18) [inline src/ArgentPonyWarcraftClient/IsExternalInit.cs:6]: Thanks for the comment on this! I remember trying to use `init` properties in a .NET Std 2.0 library and thought it was only for .NET 5. I'll definitely keep this in mind for my own work!

**Review by willwolfram18 — APPROVED** [review]
Appreciate the update to `init` for a little more backwards compatibility with existing unit tests!

**Pull request merged**